README.md
Description
===========

Installs and configures a Graylog2 server on Ubuntu (10.04 and up at present) and Centos (Tested on 6.4) systems.

This is a Chef re-engineering of the [Sean Porter][seanp] Linode [StackScript for graylog2][stackscript].

Originally developed by [Jacob Zimmerman][jbz], and taken over by [Phil Sturgeon][phil] at v0.1.0.

Recipes
=======

default
-------

Downloads, installs, configures and starts the java graylog2-server.  Does *not* install 
the web interface. Uses resources from the Opscode [apt cookbook][apt] to add a repo from which
it pulls MongoDB.

web_interface
-------------

First calls `graylog2::default` to install the server.  Then downloads, installs, and configures 
the Rails-based Graylog2 [web interface][web].  Also installs [MongoDB][mongo] and [ElasticSearch][esearch]
to support it.

apache2
-------

First calls `graylog2::web_interface` to ensure that both the server and web interface are available.  Then
calls the Opscode [apache2 cookbook][apache2] to install Apache before (itself) installing the
mod_passenger module via apt, and finally calling `apache_site` to configure Apache to serve the
web interface.


Usage
=====

This cookbook can be called via any of its three recipes.  To get a fully working graylog2 server, call
`graylog::apache2` - it will set up the full stack, with apache2 installed to serve the Rails app.  If
you use another webserver, you can call `graylog2:web_interface`, which will install the base server 
and the web interface Rails app, but not configure a webserver to serve it.

Note that this cookbook makes lots of assumptions about defaults, and does things like install a local
mongodb server with no root password (doh).  Do be sure to tweak this for your comfort and security
before using it in production.

Also note - this cookbook does *not* switch off any local logging system.  That means that if you are
trying to get Graylog2 to 'catch' syslog messages, you may have to disable your local rsyslog or 
syslog-ng as they may have already snabbled up port 514.  The cookbook *does* sudo-start Graylog2, so 
it should be able to bind udp/514 at startup.  Stop/Start/restart Graylog2 using the installed init.d
script (/etc/init.d/graylog2 stop|start|restart).  TODO: convert to an upstart job.

License and Author
==================

Author:: J.B. Zimmerman (<jzimmerman@mdsol.com>)

Copyright 2011 Medidata Solutions, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

  [apt]: http://community.opscode.com/cookbooks/apt
  [apache2]: http://community.opscode.com/cookbooks/apache2
  [mongo]: http://www.mongodb.org/
  [esearch]: http://www.elasticsearch.org/
  [jbz]: http://community.opscode.com/users/jbz
  [phil]: http://community.opscode.com/users/philsturgeon
  [seanp]: http://twitter.com/portertech
  [stackscript]: https://www.linode.com/stackscripts/view/?StackScriptID=1970
  [web]: https://github.com/Graylog2/graylog2-web-interface

CHANGELOG.md
# Changelog

## 0.13.4 (Daniel Panzella)
Changes:
* set dep to mongodb to a specific verison 0.13.4
* remove ref from berksfile for mongodb and set default to master, which would have the preoper constrained version
* removed deps in berksfile from community. Left only the non standard dep location in berksfile and defer the rest to metadata.rb
* removed metadata.json

## 0.13.4 ( Patrick McJury)
Changes:
* set dep to mongodb to a specific verison 0.13.4
* remove ref from berksfile for mongodb and set default to master, which would have the preoper constrained version
* removed deps in berksfile from community. Left only the non standard dep location in berksfile and defer the rest to metadata.rb
* removed metadata.json

test/integration/default/default_spec.rb
require 'minitest/spec'

describe_recipe 'graylog2::apache2' do
  include MiniTest::Chef::Resources
  include MiniTest::Chef::Assertions

  it "should install and start elasticsearch as a service" do
    service("elasticsearch").must_be_running
  end

  it "should install and start mongodb as a service" do
    service("mongod").must_be_running
  end
  
  it "should install and start graylog2 as a service" do
    service("graylog2").must_be_running
  end

  it "should install and start apache2 as a service" do
    service("apache2").must_be_running
  end

end

recipes/apache2.rb
#
# Cookbook Name:: graylog2
# Recipe:: apache2
#
# Copyright 2010, Medidata Solutions Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Install Graylog2 web interface
include_recipe "graylog2::web_interface"

# Install Apache using the OpsCode community cookbook
include_recipe "apache2"

#TODO Look into using passenger_apache2 cookbook for this

# Install gem dependencies
rbenv_gem "passenger" do
	ruby_version "#{node[:graylog2][:ruby_version]}"
	version "#{node[:graylog2][:passenger_version]}"
end

# Install packages to build Apache module:

node[:graylog2][:apache2][:module_build_packages].each do |name|
  package name do
    action :install
  end
end

# Build the Apache module
bash "install-apache-module" do
  cwd "#{node[:graylog2][:basedir]}/web"
  code "source /etc/profile.d/rbenv.sh && yes | passenger-install-apache2-module"
  creates "/opt/rbenv/versions/#{node[:graylog2][:ruby_version]}/lib/ruby/gems/1.9.1/gems/passenger-#{node[:graylog2][:passenger_version]}/libout/apache2/mod_passenger.so"
  user "root"
  notifies :restart, "service[apache2]"
end

# Create an Apache vhost for the Graylog2 web interface
template "apache-vhost-conf" do
  path "#{node['apache']['dir']}/sites-available/graylog2"
  source "graylog2.apache2.erb"
  mode 0644
end

# Remove the default vhost
apache_site "000-default" do
  enable false
end

# Enable the Graylog2 web interface vhost
apache_site "graylog2"

attributes/default.rb
default[:graylog2][:basedir] = "/var/graylog2"
default[:graylog2][:server][:version] = "0.12.0"
default[:graylog2][:web_interface][:version] = "0.12.0"
default[:graylog2][:web_interface][:listen_port] = 4500 # if this is not 80 you need to override default['apache']['listen_ports'] = [ "80" ] 
default[:graylog2][:web_interface][:secret_key] = "CHANGE ME"
default[:graylog2][:web_interface][:file_owner] = "nobody"
case node['platform_family']
when "rhel"
    default[:graylog2][:web_interface][:file_group] = "nobody"
when "debian"
    default[:graylog2][:web_interface][:file_group] = "nogroup"
end

case node['platform_family']
when "rhel"
    default[:graylog2][:apache2][:module_build_packages] = ["libcurl-devel","httpd-devel","apr-devel","apr-util-devel"]
when "debian"
    default[:graylog2][:apache2][:module_build_packages] = ["libcurl4-openssl-dev","apache2-threaded-dev","libapr1-dev","libaprutil1-dev"]
end

default[:rbenv][:git_repository] = "https://github.com/sstephenson/rbenv.git"
default[:ruby_build][:git_repository] = "https://github.com/sstephenson/ruby-build.git"
default[:rbenv_vars][:git_repository] = "https://github.com/sstephenson/rbenv-vars.git"

default.elasticsearch[:version]       = "0.20.4"
default.elasticsearch[:host]          = "http://download.elasticsearch.org"
default.elasticsearch[:repository]    = "elasticsearch/elasticsearch"
default.elasticsearch[:filename]      = "elasticsearch-#{node.elasticsearch[:version]}.tar.gz"
default.elasticsearch[:download_url]  = [node.elasticsearch[:host], node.elasticsearch[:repository], node.elasticsearch[:filename]].join('/')

default[:elasticsearch][:cluster][:name] = "graylog2"

default[:graylog2][:mongodb][:host] = "localhost"
default[:graylog2][:mongodb][:port] = 27017
default[:graylog2][:mongodb][:max_connections] = 150
default[:graylog2][:mongodb][:database] = "graylog2"
default[:graylog2][:mongodb][:auth] = "false"
default[:graylog2][:mongodb][:user] = "user"
default[:graylog2][:mongodb][:password] = "password"

default[:graylog2][:protocol] = "udp"
default[:graylog2][:port] = 514
default[:graylog2][:gelf_port] = 12201
default[:graylog2][:collection_size] = 50000000
default[:graylog2][:email] = "graylog2@example.org"
default[:graylog2][:configure_email] = true
default[:graylog2][:allow_deleting] = "false"
default[:graylog2][:send_stream_alarms] = true
default[:graylog2][:send_stream_subscriptions] = true
default[:graylog2][:stream_alarms_cron_minute] = "*/15"
default[:graylog2][:stream_subscriptions_cron_minute] = "*/15"

default[:graylog2][:external_hostname] = nil
default[:graylog2][:server_name] = "graylog2"

default[:graylog2][:ruby_version] = "1.9.3-p327"
default[:graylog2][:passenger_version] = "4.0.5"

default[:graylog2][:unicorn][:worker_processes] = 1
default[:graylog2][:unicorn][:timeout] = 30
default[:graylog2][:unicorn][:preload_app] = true

recipes/web_interface.rb
#
# Cookbook Name:: graylog2
# Recipe:: web_interface
#
# Copyright 2010, Medidata Solutions Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Install required APT packages
include_recipe "build-essential"

if node.graylog2.configure_email
    include_recipe "postfix"
end

# Install rbenv
include_recipe "rbenv::default"
include_recipe "rbenv::ruby_build"

# Install specific Ruby version via rbenv
rbenv_ruby "#{node.graylog2.ruby_version}"

# Install gem dependencies
%w{ bundler rake }.each do |g|
  rbenv_gem "#{g}" do
    ruby_version "#{node[:graylog2][:ruby_version]}"
  end
end

# Create the release directory
directory "#{node[:graylog2][:basedir]}/rel" do
  mode 0755
  recursive true
end

# Download the desired version of Graylog2 web interface from GitHub
remote_file "download_web_interface" do
  path "#{node[:graylog2][:basedir]}/rel/graylog2-web-interface-#{node[:graylog2][:web_interface][:version]}.tar.gz"
  source "https://github.com/Graylog2/graylog2-web-interface/releases/download/#{node[:graylog2][:web_interface][:version]}/graylog2-web-interface-#{node[:graylog2][:web_interface][:version]}.tar.gz"
  action :create_if_missing
end

# Unpack the desired version of Graylog2 web interface
execute "tar zxf graylog2-web-interface-#{node[:graylog2][:web_interface][:version]}.tar.gz" do
  cwd "#{node[:graylog2][:basedir]}/rel"
  creates "#{node[:graylog2][:basedir]}/rel/graylog2-web-interface-#{node[:graylog2][:web_interface][:version]}/build_date"
  action :nothing
  subscribes :run, resources(:remote_file => "download_web_interface"), :immediately
end

# Link to the desired Graylog2 web interface version
link "#{node[:graylog2][:basedir]}/web" do
  to "#{node[:graylog2][:basedir]}/rel/graylog2-web-interface-#{node[:graylog2][:web_interface][:version]}"
end

# Perform bundle install on the newly-installed Graylog2 web interface version
bash "bundle install" do
  cwd "#{node[:graylog2][:basedir]}/web"
  code "rbenv local #{node[:graylog2][:ruby_version]} && source /etc/profile.d/rbenv.sh && bundle install --deployment --binstubs"
  subscribes :run, resources(:link => "#{node[:graylog2][:basedir]}/web"), :immediately
end

external_hostname = node[:graylog2][:external_hostname]     ? node[:graylog2][:external_hostname] :
    (node.has_key?('ec2') and node.ec2.has_key?('public_hostname')) ? node[:ec2][:public_hostname] :
    (node.has_key?('ec2') and node.ec2.has_key?('public_ipv4'))     ? node[:ec2][:public_ipv4] :
    node.has_key?('fqdn')                                           ? node.fqdn :
    "localhost"

# Create general.yml
template "#{node.graylog2.basedir}/web/config/general.yml" do
  source "general.yml.erb"
  owner node[:graylog2][:web_interface][:file_owner]
  group node[:graylog2][:web_interface][:file_group]
  mode 0644
  variables( :external_hostname => external_hostname )
end

# Create config files
%w{ indexer ldap mongoid }.each do |yml_file|
  template "#{node[:graylog2][:basedir]}/web/config/#{yml_file}.yml" do
    source "#{yml_file}.yml.erb"
    owner node[:graylog2][:web_interface][:file_owner]
    group node[:graylog2][:web_interface][:file_group]
    mode 0644
  end
end

template "#{node.graylog2.basedir}/web/config/initializers/secret_token.rb" do
  source "secret_token.rb.erb"
  owner node[:graylog2][:web_interface][:file_owner]
  group node[:graylog2][:web_interface][:file_group]
  mode 0644
end


# Chown the Graylog2 directory to nobody/nogroup to allow web servers to serve it
execute "sudo chown -R #{node[:graylog2][:web_interface][:file_owner]}:#{node[:graylog2][:web_interface][:file_group]} graylog2-web-interface-#{node[:graylog2][:web_interface][:version]}" do
  cwd "#{node[:graylog2][:basedir]}/rel"
  not_if do
    File.stat("#{node[:graylog2][:basedir]}/rel/graylog2-web-interface-#{node[:graylog2][:web_interface][:version]}").uid == 65534
  end
  action :nothing
  subscribes :run, resources(:bash => "bundle install"), :immediately
end

# Stream message rake tasks
cron "Graylog2 send stream alarms" do
  minute node[:graylog2][:stream_alarms_cron_minute]
  action node[:graylog2][:send_stream_alarms] ? :create : :delete
  command "cd #{node[:graylog2][:basedir]}/web && RAILS_ENV=production bundle exec rake streamalarms:send"
end

cron "Graylog2 send stream subscriptions" do
  minute node[:graylog2][:stream_subscriptions_cron_minute]
  action node[:graylog2][:send_stream_subscriptions] ? :create : :delete
  command "cd #{node[:graylog2][:basedir]}/web && RAILS_ENV=production bundle exec rake subscriptions:send"
end

recipes/server.rb
#
# Cookbook Name:: graylog2
# Recipe:: server
#
# Copyright 2010, Medidata Solutions Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Install MongoDB from 10gen repository
include_recipe "mongodb::10gen_repo"
include_recipe "mongodb::default"

include_recipe "java"

# Create the release directory
directory "#{node[:graylog2][:basedir]}/rel" do
  mode 0755
  recursive true
end

include_recipe "elasticsearch"

# Download the desired version of Graylog2 server from GitHub
remote_file "download_server" do
  path "#{node[:graylog2][:basedir]}/rel/graylog2-server-#{node[:graylog2][:server][:version]}.tar.gz"
  source "https://github.com/Graylog2/graylog2-server/releases/download/#{node[:graylog2][:server][:version]}/graylog2-server-#{node[:graylog2][:server][:version]}.tar.gz"
  action :create_if_missing
end

# Unpack the desired version of Graylog2 server
execute "tar zxf graylog2-server-#{node[:graylog2][:server][:version]}.tar.gz" do
  cwd "#{node[:graylog2][:basedir]}/rel"
  creates "#{node[:graylog2][:basedir]}/rel/graylog2-server-#{node[:graylog2][:server][:version]}/build_date"
  action :nothing
  subscribes :run, resources(:remote_file => "download_server"), :immediately
end

# Link to the desired Graylog2 server version
link "#{node[:graylog2][:basedir]}/server" do
  to "#{node[:graylog2][:basedir]}/rel/graylog2-server-#{node[:graylog2][:server][:version]}"
end

# Create graylog2.conf
template "/etc/graylog2.conf" do
  mode 0644
  notifies :restart, "service[graylog2]"
end

template "/etc/graylog2-elasticsearch.yml" do
  mode 0644
  notifies :restart, "service[graylog2]"
end

# Create init.d script
template "/etc/init.d/graylog2" do
  source "graylog2.init.erb"
  mode 0755
end

# Service resource
service "mongod" do
  supports :restart => true
  action [:enable, :start]
end

# Service resource
service "graylog2" do
  supports :restart => true
  action [:enable, :start]
end
